optimum/intel/openvino/modeling_decoder.py

```diff
@@ -501,24 +506,38 @@ def forward(
                     past_key_values = tuple(
                         past_key_values[i : i + self.num_pkv] for i in range(0, len(past_key_values), self.num_pkv)
                     )
+                self.past_len += input_ids.shape[1]
             else:
                 past_key_values = None
+                self.past_len = 0
 
         return CausalLMOutputWithPast(logits=logits, past_key_values=past_key_values)
 
-    # Adapted from transformers.models.gpt2.modeling_gpt2.GPT2LMHeadModel.prepare_inputs_for_generation
+    # Adapted from transformers.models.llama.modeling_llama.LlamaForCausalLM.prepare_inputs_for_generation
     def prepare_inputs_for_generation(self, input_ids, past_key_values=None, **kwargs):
         # if model is used as a decoder in encoder-decoder model, the decoder attention mask is created on the fly
         attention_mask = kwargs.get("attention_mask", None)
         use_cache = kwargs.get("use_cache", None)
 
+        if past_key_values is not None:
+            # Keep only the unprocessed tokens:
```

Comment:
Hi, I used this PR to try speculative decoding with OV models. This PR fixes most of the problems that existed before when running with SD. I would like to suggest to add the following condition to better support SD:
```python
if past_key_values is not None:
  if not self.stateful:
    self.past_len = past_key_values[0][0].shape[2]
  ...
```
In the current implementation we remember the last sequence length the model saw but in SD the next input length is different according to the number of tokens that were accepted.

Reply:
updated